File: test/php/disbursement/BankAccountInquiryTest.php
<?php

namespace DanaUat\Disbursement;

use PHPUnit\Framework\TestCase;
use Dana\Disbursement\v1\Api\v1\Api\DisbursementApi;
use Dana\ObjectSerializer;
use Dana\Configuration;
use Dana\Env;
use Dana\ApiException;
use Dana\Disbursement\v1\Api\DisbursementApi as ApiDisbursementApi;
use DanaUat\Helper\Assertion;
use DanaUat\Helper\Util;
use Exception;

class BankAccountInquiryTest extends TestCase
{
    private static $titleCase = 'BankAccountInquiry';
    private static $jsonPathFile = 'resource/request/components/Disbursement.json';
    private static $apiInstance;
    private static $merchantId;
    
    public static function setUpBeforeClass(): void
    {
        // Set up configuration with authentication settings
        $configuration = new Configuration();

        // The Configuration constructor automatically loads values from environment variables
        // But we can manually set them if needed
        $configuration->setApiKey('PRIVATE_KEY', getenv('PRIVATE_KEY'));
        $configuration->setApiKey('ORIGIN', getenv('ORIGIN'));
        $configuration->setApiKey('X_PARTNER_ID', getenv('X_PARTNER_ID'));
        $configuration->setApiKey('ENV', Env::SANDBOX);

        // Create DisbursementApi instance directly with configuration
        self::$apiInstance = new ApiDisbursementApi(null, $configuration);

        // Store merchantId for reuse
        self::$merchantId = getenv('MERCHANT_ID');
    }

    /**
     * Should successfully inquire bank account with valid data and amount
     */
    public function testInquiryBankAccountValidDataAmount(): void
    {
        Util::withDelay(function () {
            $caseName = 'InquiryBankAccountValidDataAmount';

            try {
                // Get and prepare the request
                $jsonDict = Util::getRequest(
                    self::$jsonPathFile,
                    self::$titleCase,
                    $caseName
                );

                // Set a unique partner reference number
                $partnerReferenceNo = Util::generatePartnerReferenceNo();
                $jsonDict['partnerReferenceNo'] = $partnerReferenceNo;

                // Create a BankAccountInquiryRequest object from the JSON request data
                $bankAccountInquiryRequestObj = ObjectSerializer::deserialize(
                    $jsonDict,
                    'Dana\Disbursement\v1\Model\BankAccountInquiryRequest'
                );

                // Make the API call
                $apiResponse = self::$apiInstance->bankAccountInquiry($bankAccountInquiryRequestObj);

                // Assert the response matches the expected data
                Assertion::assertResponse(
                    self::$jsonPathFile,
                    self::$titleCase,
                    $caseName,
                    $apiResponse->__toString(),
                    ['partnerReferenceNo' => $partnerReferenceNo]
                );

                $this->assertTrue(true);
            } catch (ApiException $e) {
                $this->fail('Failed to inquire bank account: ' . $e->getMessage());
            } catch (Exception $e) {
                $this->fail('Unexpected exception: ' . $e->getMessage());
            }
        });
    }

    /**
     * Should fail to inquire bank account with insufficient fund
     */
    public function testInquiryBankAccountInsufficientFund(): void
    {
        Util::withDelay(function () {
            $caseName = 'InquiryBankAccountInsufficientFund';

            // Get and prepare the request
            $jsonDict = Util::getRequest(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName
            );

            // Set a unique partner reference number
            $partnerReferenceNo = Util::generatePartnerReferenceNo();
            $jsonDict['partnerReferenceNo'] = $partnerReferenceNo;

            // Create a BankAccountInquiryRequest object from the JSON request data
            $bankAccountInquiryRequestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Disbursement\v1\Model\BankAccountInquiryRequest'
            );

            try {
                // Make the API call
                self::$apiInstance->bankAccountInquiry($bankAccountInquiryRequestObj);

                $this->fail('Expected ApiException for insufficient fund but the API call succeeded');
            } catch (ApiException $e) {
                // We expect a 403 Forbidden for insufficient fund (responseCode: 4034214)
                $this->assertEquals(403, $e->getCode(), "Expected HTTP 403 Forbidden for insufficient fund, got {$e->getCode()}");

                // Get the response body from the exception
                $responseContent = (string)$e->getResponseBody();
                
                // Use assertFailResponse to validate the error response
                Assertion::assertFailResponse(
                    self::$jsonPathFile, 
                    self::$titleCase, 
                    $caseName, 
                    $responseContent,
                    ['partnerReferenceNo' => $partnerReferenceNo]
                );
            } catch (Exception $e) {
                $this->fail("Expected ApiException but got " . get_class($e) . ": " . $e->getMessage());
            }
        });
    }

    /**
     * Should fail to inquire bank account with unauthorized signature
     */
    public function testInquiryBankAccountUnauthorizedSignature(): void
    {
       try {
            Util::withDelay(function() {
                $caseName = 'InquiryBankAccountUnauthorizedSignature';
                $partnerReferenceNo = Util::generatePartnerReferenceNo();
                
                // Get the request data from the JSON file
                $requestData = Util::getRequest(
                    self::$jsonPathFile, 
                    self::$titleCase, 
                    $caseName
                );
                
                // Set the correct partner reference number
                $requestData['partnerReferenceNo'] = $partnerReferenceNo;
                
                // Create headers with invalid signature to test authorization failure
                $headers = Util::getHeadersWithSignature(
                    'POST', 
                    '/v1.0/emoney/bank-account-inquiry.htm',
                    $requestData,
                    true,
                    false,
                    true  // invalid_signature set to true
                );
                
                // Make direct API call with invalid signature
                try {
                    Util::executeApiRequest(
                        'POST',
                        'https://api.sandbox.dana.id/v1.0/emoney/bank-account-inquiry.htm',
                        $headers,
                        $requestData
                    );
                    
                    $this->fail('Expected ApiException for invalid signature but the API call succeeded');
                } catch (ApiException $e) {
                    // We expect a 401 Unauthorized for invalid signature
                    $this->assertEquals(401, $e->getCode(), "Expected HTTP 401 Unauthorized for invalid signature, got {$e->getCode()}");

                    // Get the response body from the exception
                    $responseContent = (string)$e->getResponseBody();
                    
                    // Use assertFailResponse to validate the error response
                    Assertion::assertFailResponse(
                        self::$jsonPathFile, 
                        self::$titleCase, 
                        $caseName, 
                        $responseContent,
                        ['partnerReferenceNo' => $partnerReferenceNo]
                    );
                } catch (Exception $e) {
                    $this->fail("Expected ApiException but got " . get_class($e) . ": " . $e->getMessage());
                }
            });
        } catch (Exception $e) {            
            $this->fail("Unexpected exception: " . $e->getMessage());
        }
    }

    /**
     * Should fail to inquire bank account with inactive account
     */
    public function testInquiryBankAccountInactiveAccount(): void
    {
        Util::withDelay(function () {
            $caseName = 'InquiryBankAccountInactiveAccount';

            // Get and prepare the request
            $jsonDict = Util::getRequest(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName
            );

            // Set a unique partner reference number
            $partnerReferenceNo = Util::generatePartnerReferenceNo();
            $jsonDict['partnerReferenceNo'] = $partnerReferenceNo;

            // Create a BankAccountInquiryRequest object from the JSON request data
            $bankAccountInquiryRequestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Disbursement\v1\Model\BankAccountInquiryRequest'
            );

            try {
                // Make the API call
                self::$apiInstance->bankAccountInquiry($bankAccountInquiryRequestObj);

                $this->fail('Expected ApiException for inactive account but the API call succeeded');
            } catch (ApiException $e) {
                // We expect a 403 Forbidden for inactive account (responseCode: 4034218)
                $this->assertEquals(403, $e->getCode(), "Expected HTTP 403 Forbidden for inactive account, got {$e->getCode()}");

                // Get the response body from the exception
                $responseContent = (string)$e->getResponseBody();
                
                // Use assertFailResponse to validate the error response
                Assertion::assertFailResponse(
                    self::$jsonPathFile, 
                    self::$titleCase, 
                    $caseName, 
                    $responseContent,
                    ['partnerReferenceNo' => $partnerReferenceNo]
                );
            } catch (Exception $e) {
                $this->fail("Expected ApiException but got " . get_class($e) . ": " . $e->getMessage());
            }
        });
    }

    /**
     * Should fail to inquire bank account with invalid merchant
     */
    public function testInquiryBankAccountInvalidMerchant(): void
    {
        Util::withDelay(function () {
            $caseName = 'InquiryBankAccountInvalidMerchant';

            // Get and prepare the request
            $jsonDict = Util::getRequest(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName
            );

            // Set a unique partner reference number
            $partnerReferenceNo = Util::generatePartnerReferenceNo();
            $jsonDict['partnerReferenceNo'] = $partnerReferenceNo;

            // Create a BankAccountInquiryRequest object from the JSON request data
            $bankAccountInquiryRequestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Disbursement\v1\Model\BankAccountInquiryRequest'
            );

            try {
                // Make the API call
                self::$apiInstance->bankAccountInquiry($bankAccountInquiryRequestObj);

                $this->fail('Expected ApiException for invalid merchant but the API call succeeded');
            } catch (ApiException $e) {
                // We expect a 404 Not Found for invalid merchant (responseCode: 4044208)
                $this->assertEquals(404, $e->getCode(), "Expected HTTP 404 Not Found for invalid merchant, got {$e->getCode()}");

                // Get the response body from the exception
                $responseContent = (string)$e->getResponseBody();
                
                // Use assertFailResponse to validate the error response
                Assertion::assertFailResponse(
                    self::$jsonPathFile, 
                    self::$titleCase, 
                    $caseName, 
                    $responseContent,
                    ['partnerReferenceNo' => $partnerReferenceNo]
                );
            } catch (Exception $e) {
                $this->fail("Expected ApiException but got " . get_class($e) . ": " . $e->getMessage());
            }
        });
    }

    /**
     * Should fail to inquire bank account with invalid card
     */
    public function testInquiryBankAccountInvalidCard(): void
    {
        Util::withDelay(function () {
            $caseName = 'InquiryBankAccountInvalidCard';

            // Get and prepare the request
            $jsonDict = Util::getRequest(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName
            );

            // Set a unique partner reference number
            $partnerReferenceNo = Util::generatePartnerReferenceNo();
            $jsonDict['partnerReferenceNo'] = $partnerReferenceNo;

            // Create a BankAccountInquiryRequest object from the JSON request data
            $bankAccountInquiryRequestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Disbursement\v1\Model\BankAccountInquiryRequest'
            );

            try {
                // Make the API call
                self::$apiInstance->bankAccountInquiry($bankAccountInquiryRequestObj);

                $this->fail('Expected ApiException for invalid card but the API call succeeded');
            } catch (ApiException $e) {
                // We expect a 404 Not Found for invalid card (responseCode: 4044211)
                $this->assertEquals(404, $e->getCode(), "Expected HTTP 404 Not Found for invalid card, got {$e->getCode()}");

                // Get the response body from the exception
                $responseContent = (string)$e->getResponseBody();
                
                // Use assertFailResponse to validate the error response
                Assertion::assertFailResponse(
                    self::$jsonPathFile, 
                    self::$titleCase, 
                    $caseName, 
                    $responseContent,
                    ['partnerReferenceNo' => $partnerReferenceNo]
                );
            } catch (Exception $e) {
                $this->fail("Expected ApiException but got " . get_class($e) . ": " . $e->getMessage());
            }
        });
    }

    /**
     * Should fail to inquire bank account with invalid field format
     */
    public function testInquiryBankAccountInvalidFieldFormat(): void
    {
        Util::withDelay(function () {
            $caseName = 'InquiryBankAccountInvalidFieldFormat';

            // Get and prepare the request
            $jsonDict = Util::getRequest(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName
            );

            // Set a unique partner reference number
            $partnerReferenceNo = Util::generatePartnerReferenceNo();
            $jsonDict['partnerReferenceNo'] = $partnerReferenceNo;

            // Create a BankAccountInquiryRequest object from the JSON request data
            $bankAccountInquiryRequestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Disbursement\v1\Model\BankAccountInquiryRequest'
            );

            try {
                // Make the API call
                self::$apiInstance->bankAccountInquiry($bankAccountInquiryRequestObj);

                $this->fail('Expected ApiException for invalid field format but the API call succeeded');
            } catch (ApiException $e) {
                // We expect a 400 Bad Request for invalid field format (responseCode: 4004201)
                $this->assertEquals(400, $e->getCode(), "Expected HTTP 400 Bad Request for invalid field format, got {$e->getCode()}");

                // Get the response body from the exception
                $responseContent = (string)$e->getResponseBody();
                
                // Use assertFailResponse to validate the error response
                Assertion::assertFailResponse(
                    self::$jsonPathFile, 
                    self::$titleCase, 
                    $caseName, 
                    $responseContent,
                    ['partnerReferenceNo' => $partnerReferenceNo]
                );
            } catch (Exception $e) {
                $this->fail("Expected ApiException but got " . get_class($e) . ": " . $e->getMessage());
            }
        });
    }

    /**
     * Should fail to inquire bank account with missing mandatory field
     */
    public function testInquiryBankAccountMissingMandatoryField(): void
    {
        Util::withDelay(function () {
            $caseName = 'InquiryBankAccountMissingMandatoryField';

            // Get and prepare the request
            $jsonDict = Util::getRequest(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName
            );

            // Set a unique partner reference number
            $partnerReferenceNo = Util::generatePartnerReferenceNo();
            $jsonDict['partnerReferenceNo'] = $partnerReferenceNo;

            // Create a BankAccountInquiryRequest object from the JSON request data
            $bankAccountInquiryRequestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Disbursement\v1\Model\BankAccountInquiryRequest'
            );

            try {
                // Make the API call
                self::$apiInstance->bankAccountInquiry($bankAccountInquiryRequestObj);

                $this->fail('Expected ApiException for missing mandatory field but the API call succeeded');
            } catch (ApiException $e) {
                // We expect a 400 Bad Request for missing mandatory field (responseCode: 4004202)
                $this->assertEquals(400, $e->getCode(), "Expected HTTP 400 Bad Request for missing mandatory field, got {$e->getCode()}");

                // Get the response body from the exception
                $responseContent = (string)$e->getResponseBody();
                
                // Use assertFailResponse to validate the error response
                Assertion::assertFailResponse(
                    self::$jsonPathFile, 
                    self::$titleCase, 
                    $caseName, 
                    $responseContent,
                    ['partnerReferenceNo' => $partnerReferenceNo]
                );
            } catch (Exception $e) {
                $this->fail("Expected ApiException but got " . get_class($e) . ": " . $e->getMessage());
            }
        });
    }
}

File: test/python/disbursement/bank_account_inquiry_test.py
import os
import pytest
from dana.utils.snap_configuration import SnapConfiguration, AuthSettings, Env
from dana.disbursement.v1.enum import *
from dana.disbursement.v1.models import *
from dana.disbursement.v1 import *
from dana.disbursement.v1.api import *
from dana.api_client import ApiClient
from dana.exceptions import *
from uuid import uuid4
from helper.util import get_request, with_delay, retry_on_inconsistent_request
from helper.api_helpers import get_headers_with_signature, execute_and_assert_api_error
from helper.assertion import assert_response, assert_fail_response

title_case = "BankAccountInquiry"
json_path_file = "resource/request/components/Disbursement.json"
BANK_ACCOUNT_INQUIRY_ENDPOINT = "https://api.sandbox.dana.id/v1.0/emoney/bank-account-inquiry.htm"
BANK_ACCOUNT_INQUIRY_RESOURCE_PATH = "/v1.0/emoney/bank-account-inquiry.htm"

configuration = SnapConfiguration(
    api_key=AuthSettings(
        PRIVATE_KEY=os.environ.get("PRIVATE_KEY"),
        ORIGIN=os.environ.get("ORIGIN"),
        X_PARTNER_ID=os.environ.get("X_PARTNER_ID"),
        DANA_ENV=Env.SANDBOX,
        X_DEBUG="true",
        CLIENT_SECRET=os.environ.get("CLIENT_SECRET"),
    )
)

with ApiClient(configuration) as api_client:
    api_instance = DisbursementApi(api_client)


def _assert_bank_account_inquiry_error_bypass_sdk(case_name: str, expected_status: int) -> None:
    """Hit the API directly so sandbox SDK validation does not block error fixtures."""
    json_dict = get_request(json_path_file, title_case, case_name)
    partner_reference_no = str(uuid4())
    json_dict["partnerReferenceNo"] = partner_reference_no

    headers = get_headers_with_signature(
        method="POST",
        resource_path=BANK_ACCOUNT_INQUIRY_RESOURCE_PATH,
        request_obj=json_dict,
        with_timestamp=True,
        invalid_timestamp=False,
    )

    execute_and_assert_api_error(
        api_client,
        "POST",
        BANK_ACCOUNT_INQUIRY_ENDPOINT,
        json_dict,
        headers,
        expected_status,
        json_path_file,
        title_case,
        case_name,
        {"partnerReferenceNo": partner_reference_no},
    )

@with_delay()
@retry_on_inconsistent_request(max_retries=3, delay_seconds=2)
def test_inquiry_bank_account_valid_data_amount():
    case_name = "InquiryBankAccountValidDataAmount"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    json_dict["partnerReferenceNo"] = str(uuid4())
    request_obj = BankAccountInquiryRequest.from_dict(json_dict)

    try:
        response = api_instance.bank_account_inquiry(request_obj)
        assert_response(json_path_file, title_case, case_name, BankAccountInquiryResponse.to_json(response), {"partnerReferenceNo": json_dict["partnerReferenceNo"]})

    except ApiException as e:
        pytest.fail(f"API call failed: {e}")

@with_delay()
@retry_on_inconsistent_request(max_retries=3, delay_seconds=2)
def test_inquiry_bank_account_insufficient_fund():
    _assert_bank_account_inquiry_error_bypass_sdk("InquiryBankAccountInsufficientFund", 403)

@with_delay()
@retry_on_inconsistent_request(max_retries=3, delay_seconds=2)
def test_inquiry_bank_account_inactive_account():
    _assert_bank_account_inquiry_error_bypass_sdk("InquiryBankAccountInactiveAccount", 403)

@with_delay()
@retry_on_inconsistent_request(max_retries=3, delay_seconds=2)
def test_inquiry_bank_account_invalid_merchant():
    _assert_bank_account_inquiry_error_bypass_sdk("InquiryBankAccountInvalidMerchant", 404)

@with_delay()
@retry_on_inconsistent_request(max_retries=3, delay_seconds=2)
def test_inquiry_bank_account_invalid_card():
    _assert_bank_account_inquiry_error_bypass_sdk("InquiryBankAccountInvalidCard", 404)

@with_delay()
@retry_on_inconsistent_request(max_retries=3, delay_seconds=2)
def test_inquiry_bank_account_invalid_field_format():
    case_name = "InquiryBankAccountInvalidFieldFormat"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    json_dict["partnerReferenceNo"] = str(uuid4())
    request_obj = BankAccountInquiryRequest.from_dict(json_dict)

    try:
        response = api_instance.bank_account_inquiry(request_obj)
        assert_fail_response(json_path_file, title_case, case_name, BankAccountInquiryResponse.to_json(response), 
                           {'partnerReferenceNo': json_dict["partnerReferenceNo"]})

    except ApiException as e:
        assert_fail_response(json_path_file, title_case, case_name, str(e.body), 
                           {'partnerReferenceNo': json_dict["partnerReferenceNo"]})

@with_delay()
@retry_on_inconsistent_request(max_retries=3, delay_seconds=2)
def test_inquiry_bank_account_missing_mandatory_field():
    _assert_bank_account_inquiry_error_bypass_sdk("InquiryBankAccountMissingMandatoryField", 400)

@with_delay()
@retry_on_inconsistent_request(max_retries=3, delay_seconds=2)
def test_inquiry_bank_account_unauthorized_signature():
    case_name = "InquiryBankAccountUnauthorizedSignature"
    json_dict = get_request(json_path_file, title_case, case_name)
    
    test_order_reference_number = str(uuid4())
    json_dict["partnerReferenceNo"] = test_order_reference_number
    request_obj = BankAccountInquiryRequest.from_dict(json_dict)

    headers = get_headers_with_signature(
        method="POST",
        resource_path=BANK_ACCOUNT_INQUIRY_RESOURCE_PATH,
        request_obj=json_dict,
        with_timestamp=True,
        invalid_timestamp=False
    )
    
    headers["X-SIGNATURE"] = "85be817c55b2c135157c7e89f52499bf0c25ad6eeebe04a986e8c862561b19a5"

    execute_and_assert_api_error(
        api_client,
        "POST",
        BANK_ACCOUNT_INQUIRY_ENDPOINT,
        request_obj,
        headers,
        401,
        json_path_file,
        title_case,
        case_name,
        {"partnerReferenceNo": test_order_reference_number}
    )
